policybot/pkg/testflakes/testflakes.go

```diff
@@ -0,0 +1,358 @@
+	bucket := client.Bucket("istio-prow")
+
+	var allTestRuns = []*store.TestFlake{}
+
+	for _, test := range testSlice {
+		testName := test.Name
+
+		prefs := test.Prs
+
+		for _, pref := range prefs {
+
+			var onePull = &store.TestFlake{}
+
+			onePull.TestName = testName
+			onePull.RunPath = pref
+
+			obj := bucket.Object(pref + "/clone-records.json")
+			log.Println("read clone")
+			rdr, err := obj.NewReader(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			defer rdr.Close()
+			slurp, err := ioutil.ReadAll(rdr)
+			if err != nil {
+				return nil, err
+			}
+			s := string(slurp)
+			dec := json.NewDecoder(strings.NewReader(s))
+
+			t, err := dec.Token()
+			if err != nil {
+				return nil, err
+			}
+			fmt.Printf("%T: %v\n", t, t)
+
+			for dec.More() {
+				var record CloneRecord
+				err := dec.Decode(&record)
+				if err != nil {
+					return nil, err
+				}
+
+				refs := record.Refs
+				pulls := refs.Pulls
+				pull := pulls[0]
+				sha := pull.Sha
+				baseSha := refs.BaseSha
+
+				failed := record.Failed
+
+				onePull.Sha = sha
+				onePull.BaseSha = baseSha
+				onePull.CloneFailed = failed
+
+			}
+			newObj := bucket.Object(pref + "/started.json")
+			nrdr, nerr := newObj.NewReader(ctx)
+			if nerr != nil {
+				return nil, nerr
+			}
+
+			defer nrdr.Close()
+			slur, nerr := ioutil.ReadAll(nrdr)
+			if nerr != nil {
+				return nil, nerr
+			}
+			ns := string(slur)
+			ndec := json.NewDecoder(strings.NewReader(ns))
+
+			for ndec.More() {
+				var started Started
+				err = ndec.Decode(&started)
+				if err != nil {
+					return nil, err
+				}
+
+				t := started.Timestamp
+				tm := time.Unix(t, 0)
+				onePull.StartTime = tm
+
+			}
+
+			// It is possible that the folder might not contain finished.json.
+			newObj = bucket.Object(pref + "/finished.json")
+			nrdr, nerr = newObj.NewReader(ctx)
+			if nerr != nil {
+				return nil, nerr
+			}
+
+			defer nrdr.Close()
+			slur, nerr = ioutil.ReadAll(nrdr)
+			if err != nil {
+				return nil, nerr
+			}
+			ns = string(slur)
+			ndec = json.NewDecoder(strings.NewReader(ns))
+
+			for ndec.More() {
+				var finished Finished
+				err = ndec.Decode(&finished)
+				if err != nil {
+					return nil, err
+				}
+
+				passed := finished.Passed
+				result := finished.Result
+				t := finished.Timestamp
+				tm := time.Unix(t, 0)
+
+				onePull.TestPassed = passed
+				onePull.Result = result
+				onePull.FinishTime = tm
+			}
+
+			prefSplit := strings.Split(pref, "/")
+
+			runNo, errr := strconv.ParseInt(prefSplit[len(prefSplit)-1], 10, 64)
+			if errr != nil {
+				return nil, errr
+			}
+			onePull.RunNum = runNo
+			prNo, newError := strconv.ParseInt(prefSplit[len(prefSplit)-3], 10, 64)
+			if newError != nil {
+				return nil, newError
+			}
+			onePull.PrNum = prNo
+			allTestRuns = append(allTestRuns, onePull)
+
+		}
+	}
+	return allTestRuns, nil
+}
+
+func (prFlakeTest PrFlakeTest) SetOrgID(orgID string, testFlakes []*store.TestFlake) []*store.TestFlake {
```

Comment:
In general, it's much preferable to have immutable objects when you can. I think the org id should be set at the time that the flake are observed, rather than after the fact.